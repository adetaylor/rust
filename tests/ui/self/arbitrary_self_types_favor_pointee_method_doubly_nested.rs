// run-pass

#![feature(receiver_trait)]
#![allow(dead_code)]

struct Foo;
struct InnerPtr<T>(T);
struct OuterPtr<T>(T);

impl<T> std::ops::Receiver for InnerPtr<T> {
    type Target = T;
}

impl<T> std::ops::Receiver for OuterPtr<T> {
    type Target = T;
}

impl Foo {
    fn m(self: &OuterPtr<InnerPtr<Self>>) -> u32 { 1 }
}

impl<T> InnerPtr<T> {
    fn m(self: &OuterPtr<Self>) -> u32 { 2 }
}

impl<T> OuterPtr<T> {
    fn m(self) -> u32 { 3 }
}

fn main() {
    let  a = OuterPtr(InnerPtr(Foo));
    assert_eq!(a.m(), 1);
    //~^ ambiguous target
}
