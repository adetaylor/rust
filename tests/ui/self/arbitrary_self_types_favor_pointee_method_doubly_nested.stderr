warning: InnerPtr::<T>::m and Foo::m are ambiguous targets of this method call; explicitly choose one
  --> $DIR/arbitrary_self_types_favor_pointee_method_doubly_nested.rs:32:18
   |
LL |     fn m(self: &OuterPtr<InnerPtr<Self>>) -> u32 { 1 }
   |     -------------------------------------------- because we'll call this method
...
LL |     fn m(self: &OuterPtr<Self>) -> u32 { 2 }
   |     ---------------------------------- this method won't be called
...
LL |     assert_eq!(a.m(), 1);
   |                  ^
   |
   = note: `#[warn(shadowed_method_picked)]` on by default
help: call as a function not a method
   |
LL |     assert_eq!(a.Foo::m(FOO, ARGUMENTS)(), 1);
   |                  ~~~~~~~~~~~~~~~~~~~~~~
help: call as a function not a method
   |
LL |     assert_eq!(a.InnerPtr::<T>::m(FOO, ARGUMENTS)(), 1);
   |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

warning: OuterPtr::<T>::m and Foo::m are ambiguous targets of this method call; explicitly choose one
  --> $DIR/arbitrary_self_types_favor_pointee_method_doubly_nested.rs:32:18
   |
LL |     fn m(self: &OuterPtr<InnerPtr<Self>>) -> u32 { 1 }
   |     -------------------------------------------- because we'll call this method
...
LL |     fn m(self) -> u32 { 3 }
   |     ----------------- this method won't be called
...
LL |     assert_eq!(a.m(), 1);
   |                  ^
   |
help: call as a function not a method
   |
LL |     assert_eq!(a.Foo::m(FOO, ARGUMENTS)(), 1);
   |                  ~~~~~~~~~~~~~~~~~~~~~~
help: call as a function not a method
   |
LL |     assert_eq!(a.OuterPtr::<T>::m(FOO, ARGUMENTS)(), 1);
   |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

warning: 2 warnings emitted

